src/app/AttributePathExpandIterator-DataModel.cpp

```diff
@@ -0,0 +1,242 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "lib/support/CodeUtils.h"
+#include <app/AttributePathExpandIterator-DataModel.h>
+#include <app/GlobalAttributes.h>
+
+using namespace chip::app::InteractionModel;
+
+namespace chip {
+namespace app {
+
+AttributePathExpandIteratorDataModel::AttributePathExpandIteratorDataModel(
+    InteractionModel::DataModel * dataModel, SingleLinkedListNode<AttributePathParams> * attributePath) :
+    mDataModel(dataModel),
+    mpAttributePath(attributePath), mOutputPath(kInvalidEndpointId, kInvalidClusterId, kInvalidAttributeId)
+
+{
+    mOutputPath.mExpanded = true; // this is reset in 'next' if needed
+
+    // Make the iterator ready to emit the first valid path in the list.
+    // TODO: the bool return value here is completely unchecked
+    Next();
+}
+
+bool AttributePathExpandIteratorDataModel::IsValidAttributeId(AttributeId attributeId)
+{
+    switch (attributeId)
+    {
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id:
+    case Clusters::Globals::Attributes::AcceptedCommandList::Id:
+#if CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+    case Clusters::Globals::Attributes::EventList::Id:
+#endif // CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+    case Clusters::Globals::Attributes::AttributeList::Id:
+        return true;
+    default:
+        break;
+    }
+
+    const ConcreteAttributePath attributePath(mOutputPath.mEndpointId, mOutputPath.mClusterId, attributeId);
+    return mDataModel->GetAttributeInfo(attributePath).has_value();
+}
+
+std::optional<AttributeId> AttributePathExpandIteratorDataModel::NextAttributeId()
+{
+    if (mOutputPath.mAttributeId == kInvalidAttributeId)
+    {
+        if (mpAttributePath->mValue.HasWildcardAttributeId())
+        {
+            AttributeEntry entry = mDataModel->FirstAttribute(mOutputPath);
+            return entry.IsValid()                                         //
+                ? entry.path.mAttributeId                                  //
+                : Clusters::Globals::Attributes::GeneratedCommandList::Id; //
+        }
+
+        // We allow fixed attribute IDs if and only if they are valid:
+        //    - they may be GLOBAL attributes OR
+        //    - they are valid attributes for this cluster
+        if (IsValidAttributeId(mpAttributePath->mValue.mAttributeId))
+        {
+            return mpAttributePath->mValue.mAttributeId;
+        }
+
+        return std::nullopt;
+    }
+
+    // advance the existing attribute id if it can be advanced
+    VerifyOrReturnValue(mpAttributePath->mValue.HasWildcardAttributeId(), std::nullopt);
+
+    // Ensure (including ordering) that GlobalAttributesNotInMetadata is reported as needed
+    for (unsigned i = 0; i < ArraySize(GlobalAttributesNotInMetadata); i++)
```

Comment:
I am unsure if it makes a difference. I would prefer not to touch it because that preserves CI runs so far and also I believe `int/unsigned` tells the compiler to just use whatever is most efficient. We likely do not specifically need a size for this loop.